### PR TITLE
Feature/127 like challenge

### PR DESCRIPTION
**Pull request by Limbo2332**
https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/93576084/587f2f51-366b-4bbc-a1a4-4876cf34c01a

### CODE REVIEW

File: backend/LeetWars.Core/LeetWars.Core.BLL/MappingProfiles/ChallengeProfile.cs [hunk 1]
```diff
@@ -29,6 +34,11 @@ public ChallengeProfile()
             CreateMap<ChallengeLevel, ChallengeLevelDto>();
             
             CreateMap<Challenge, ChallengeFullDto>();
+
+            CreateMap<ChallengeStarDto, ChallengeStar>()
+                .ForMember(dest => dest.ChallengeId, opt => opt.MapFrom(src => src.Challenge.Id))
+                .ForMember(dest => dest.Challenge, opt => opt.Ignore())
+                .ForMember(dest => dest.Author, opt => opt.Ignore());
```
Comment (SashaBondarchuk) [inline backend/LeetWars.Core/LeetWars.Core.BLL/MappingProfiles/ChallengeProfile.cs]: Ignore for author can be removed

File: frontend/src/app/modules/main/challenge/challenge.component.ts [hunk 2]
```diff
@@ -10,9 +10,15 @@ import { getLanguageIconUrl } from '@shared/utils/language-icons';
...
+
     public challengeStars = 0;
 
     public challengePositiveFeedbacksPercent = 0;
 
     public getLanguageIconUrl = getLanguageIconUrl;
+
+    public starChange() {
+        this.starChallenge.emit(this.challenge);
+    }
```
Comment (SashaBondarchuk) [inline frontend/src/app/modules/main/challenge/challenge.component.ts]: you can move out logic with starService here from filtering section component and avoid using Output

File: frontend/src/app/modules/main/filtering-section/filtering-section.component.html [hunk 3]
```diff
@@ -43,7 +43,7 @@ <h4>No available challenge</h4>
         </div>
 
         <div *ngFor="let challenge of challenges">
-            <app-challenge [challenge]="challenge"></app-challenge>
+            <app-challenge [challenge]="challenge" (starChallenge)="updateChallengeStar($event)"></app-challenge>
```
Comment (SashaBondarchuk) [inline frontend/src/app/modules/main/filtering-section/filtering-section.component.html]: see previous comment

File: frontend/src/app/modules/main/filtering-section/filtering-section.component.ts [hunk 4]
```diff
@@ -113,13 +125,26 @@ export class FilteringSectionComponent extends BaseComponent implements OnInit {
...
+            challenge,
+            isStar: !challenge.isStarry,
+        };
+
+        this.starService.updateStar(star).subscribe((starResp: Star) => {
+            challenge.isStarry = starResp.isStar;
+            challenge.starsAmount = starResp.isStar ? challenge.starsAmount + 1 : challenge.starsAmount - 1;
+        });
+    }
+
```
Comment (SashaBondarchuk) [inline frontend/src/app/modules/main/filtering-section/filtering-section.component.ts]: see previous comment

**Review by SashaBondarchuk — CHANGES_REQUESTED** [review]

**Review by SashaBondarchuk — APPROVED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Interfaces/IStarService.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using LeetWars.Core.Common.DTO.ChallengeStar;
+
+namespace LeetWars.Core.BLL.Interfaces
+{
+    public interface IStarService
+    {
+        Task<ChallengeStarDto> Update(ChallengeStarDto challengeStarDto);
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Interfaces/IStarService.cs]: do we really need a separate service for this?

**Comment by Limbo2332** [comment]
[Like the challenge](https://trello.com/c/lqXepjTk/127-like-the-challenge)

File: frontend/src/app/modules/main/challenge/challenge.component.ts [hunk 6]
```diff
@@ -8,11 +12,32 @@ import { getLanguageIconUrl } from '@shared/utils/language-icons';
...
+            authorId: this.user.id,
+            challenge: this.challenge,
+            isStar: !this.challenge.isStarry,
+        };
+
+        this.starService.updateStar(star).subscribe((starResp: Star) => {
+            this.challenge.isStarry = starResp.isStar;
+            this.challenge.starsAmount = starResp.isStar
+                ? this.challenge.starsAmount + 1
+                : this.challenge.starsAmount - 1;
```
Comment (tatianahutii) [inline frontend/src/app/modules/main/challenge/challenge.component.ts]: maybe it would be better to reload challenge?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Interfaces/IStarService.cs (hunk 5 above)
Comment (Limbo2332) [reply backend/LeetWars.Core/LeetWars.Core.BLL/Interfaces/IStarService.cs]: I just thought that we update stars not challenges, but now, after refactoring, it really looks like we change challenge, so we don't need starservice anymore

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**